dcap/types/src/collateral.rs

```
@@ -0,0 +1,805 @@
+//      verification collateral must support the new 3.1 APIs.
+//
+// This version can be ensured by setting the `pccs_api_version` key of `/etc/sgx_default_qcnl.conf`
+// to `3.1`.
+//
+//      ,"pccs_api_version": "3.1"
+//
+const VERSION_MAJOR: u16 = 3;
+const VERSION_MINOR: u16 = 1;
+
+/// Error creating a [`Collateral`]
+#[derive(Debug, displaydoc::Display, Clone, Eq, PartialEq)]
+pub enum Error {
+    /// Error converting from DER {0}
+    Der(x509_cert::der::Error),
+    /// Unsupported version, expected 3.1, but got {0}.{1}
+    Version(u16, u16),
+    /// Unsupported collateral service, expected SGX (0), but got {0}
+    CollateralService(u32),
+    /// Error converting bytes to String {0}
+    Utf8(FromUtf8Error),
+    /// One of the collateral fields is missing
+    MissingCollateral,
+}
+
+impl From<x509_cert::der::Error> for Error {
+    fn from(err: x509_cert::der::Error) -> Self {
+        Self::Der(err)
+    }
+}
+
+impl From<FromUtf8Error> for Error {
+    fn from(err: FromUtf8Error) -> Self {
+        Self::Utf8(err)
+    }
+}
+
+/// Collateral for a quote.
+///
+/// This data can be retrieved from the individual endpoints:
+/// - <https://api.portal.trustedservices.intel.com/documentation#pcs-revocation-v4>
+/// - <https://api.portal.trustedservices.intel.com/documentation#pcs-tcb-info-v4>
+/// - <https://api.portal.trustedservices.intel.com/documentation#pcs-enclave-identity-v4>
+/// - <https://certificates.trustedservices.intel.com/IntelSGXRootCA.der>
+///
+/// or the data can be retrieved from `sgx_ql_get_quote_verification_collateral()`
+///
+/// The certificate chains and CRLs are documented in
+/// <https://api.trustedservices.intel.com/documents/Intel_SGX_PCK_Certificate_CRL_Spec-1.5.pdf>
+#[derive(Debug, Clone)]
+pub struct Collateral {
+    root_ca_crl: CertificateList,
+    pck_crl_issuer_chain: Vec<Certificate>,
+    pck_crl: CertificateList,
+    tcb_issuer_chain: Vec<Certificate>,
+    tcb_info: String,
```

This String is what would be fed into https://github.com/mobilecoinfoundation/attestation/blob/945015c9cf6510766fddf8ecfe383cd5e29fc161/verifier/src/tcb.rs
I think perhaps we should consider moving that module into here and this should be a TCB type, which propagates an error out if the underlying str is invalid.